toolchain/check/import_cpp.cpp

```diff
@@ -77,6 +79,80 @@ auto ImportCppFiles(
     context.emitter().Emit(loc, CppInteropParseWarning, num_warnings,
                            num_imports, diagnostics_str);
   }
+  return ast;
+}
+
+// Adds a namespace for the `Cpp` import and returns its `NameScopeId`.
+static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
+                         llvm::ArrayRef<Parse::Tree::PackagingNames> imports)
+    -> SemIR::NameScopeId {
+  // TODO: Extract and deudplicate common logic with import.cpp.
+  auto name_id = SemIR::NameId::ForIdentifier(cpp_package_id);
+
+  Parse::ImportDeclId first_import_node_id = imports.front().node_id;
+
+  SemIR::InstId import_cpp_id =
+      context.AddInst<SemIR::ImportCppDecl>(first_import_node_id, {});
+
+  auto& import_cpps = context.sem_ir().import_cpps();
+  import_cpps.Reserve(imports.size());
+  for (const Parse::Tree::PackagingNames& import : imports) {
+    import_cpps.Add(
+        {.node_id = import.node_id, .library_id = import.library_id});
+  }
+
+  auto namespace_inst = SemIR::Namespace{
+      .type_id =
+          context.GetSingletonType(SemIR::NamespaceType::SingletonInstId),
+      .name_scope_id = SemIR::NameScopeId::None,
+      .import_id = import_cpp_id};
+
+  auto namespace_inst_and_loc =
+      // TODO: Check that this actually is an `AnyNamespaceId`.
+      SemIR::LocIdAndInst(Parse::AnyNamespaceId(first_import_node_id),
+                          namespace_inst);
+
+  SemIR::InstId namespace_id =
+      context.AddPlaceholderInstInNoBlock(namespace_inst_and_loc);
+  context.import_ref_ids().push_back(namespace_id);
+  SemIR::NameScope& package_name_scope =
+      context.name_scopes().Get(SemIR::NameScopeId::Package);
+  auto [added, entry_id] = package_name_scope.LookupOrAdd(
+      name_id, namespace_id, SemIR::AccessKind::Public);
+  CARBON_CHECK(added ||
+               !package_name_scope.GetEntry(entry_id).result.is_poisoned());
+
+  namespace_inst.name_scope_id = context.name_scopes().Add(
+      namespace_id, name_id, SemIR::NameScopeId::Package);
+  context.ReplaceInstBeforeConstantUse(namespace_id, namespace_inst);
+
+  context.name_scopes()
+      .Get(namespace_inst.name_scope_id)
+      .set_is_closed_import(true);
+
+  return namespace_inst.name_scope_id;
+}
+
+auto ImportCppFiles(Context& context, llvm::StringRef importing_file_path,
+                    llvm::ArrayRef<Parse::Tree::PackagingNames> imports,
+                    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs)
+    -> void {
+  if (imports.empty()) {
+    return;
+  }
+
+  auto ast = GenerateAst(context, importing_file_path, imports, fs);
+
+  IdentifierId package_id = imports.front().package_id;
+  CARBON_CHECK(
+      llvm::all_of(imports, [&](const Parse::Tree::PackagingNames& import) {
+        return import.package_id == package_id;
+      }));
+  auto name_scope_id = AddNamespace(context, package_id, imports);
+
+  if (!ast) {
+    context.name_scopes().Get(name_scope_id).set_has_error();
+  }
```

Comment:
I think the separate changes to `GenerateAST` mean an `ast` will always be non-null here. Note though, you could make `GenerateAst` return something like `std::pair<Ast, bool has_error>`. So I think you probably want one of:

```suggestion
```

```suggestion

  if (ast_has_error) {
    context.name_scopes().Get(name_scope_id).set_has_error();
  }
```

Reply:
Done.
Note that `ast` could be null, but not due to errors.

Reply:
Sorry, I'm definitely missing this Can you document the return value of `GenerateAst`, both to explain the `bool` and to explain when the AST will be null?

Reply:
Digging further, it looks like you're talking about buildASTFromCodeWithArgs can return null if Invocation.run returns false. I wonder if this makes the `has_errors` check redundant right now... but also, maybe it means that we'll want to make our own version that returns an AST even when there are errors?

To be sure, not asking for changes to that here, just trying to make sure my understanding is correct.

Reply:
Elaborated on the return value and added a TODO.
I believe that we should re-evaluate once we have a serialized AST.